docs/extending/extending_draftail.md

```diff
@@ -329,17 +325,203 @@ To fully complete the demo, we can add a bit of JavaScript to the front-end in o
 
+## Other editor extensions
+
+Draftail has additional APIs for more arbitrary customisations:
+
+- **Controls** – To add arbitrary UI elements to editor toolbars.
+- **Decorators** – For arbitrary text decorations / highlighting.
+- **Plugins** – For direct access to all Draft.js APIs.
+
+### Custom toolbar controls
+
+To add an arbitrary new UI element to editor toolbars, Draftail comes with a [controls API](https://www.draftail.org/docs/arbitrary-controls). Controls can be arbitrary React components, which can get and set the editor state. Note controls update on _every keystroke_ in the editor – make sure they render fast!
+
+Here is an example with a simple sentence counter – first, registering the editor feature in a `wagtail_hooks.py`:
+
+```python
+from wagtail.admin.rich_text.editors.draftail.features import ControlFeature
+
+
+@hooks.register('register_rich_text_features')
+def register_sentences_counter(features):
+    feature_name = 'sentences'
+    features.default_features.append(feature_name)
+
+    features.register_editor_plugin(
+        'draftail',
+        feature_name,
+        ControlFeature({
+            'type': feature_name,
+        },
+        js=['draftail_sentences.js'],
+        ),
+    )
+```
+
+Then, `draftail_sentences.js` declares a React component that will be rendered in the "meta" bottom toolbar of the editor:
+
+```javascript
+const countSentences = (str) =>
+  str ? (str.match(/[.?!…]+./g) || []).length + 1 : 0;
+
+const SentenceCounter = ({ getEditorState }) => {
+  const editorState = getEditorState();
+  const content = editorState.getCurrentContent();
+  const text = content.getPlainText();
+
+  return window.React.createElement('div', {
+    className: 'w-inline-block w-tabular-nums w-help-text w-mr-4',
+  }, `Sentences: ${countSentences(text)}`);
+}
+
+window.draftail.registerPlugin({
+  type: 'sentences',
+  meta: SentenceCounter,
+}, 'controls');
+```
+
+### Text decorators
+
+The [decorators API](https://www.draftail.org/docs/decorators) is how Draftail / Draft.js supports highlighting text with special formatting in the editor. It uses the [CompositeDecorator](https://draftjs.org/docs/advanced-topics-decorators/#compositedecorator) API, with each entry having a `strategy` function to determine what text to target, and a `component` function to render the decoration.
+
+There are two important considerations when using this API:
+
+- Order matters: only one decorator can render per character in the editor. This includes any entities that are rendered as decorations.
+- For performance reasons, Draft.js only re-renders decorators that are on the currently-focused line of text.
+
+Here is an example with highlighting of problematic punctuation – first, registering the editor feature in a `wagtail_hooks.py`:
```

Comment:
I chose this example because it’s language-agnostic and the logic on what to decorate is simple.